tasty-plutus/src/Test/Tasty/Plutus/Internal/TestScript.hs

```diff
@@ -0,0 +1,288 @@
+ the datum and redeemer types of the 'Validator' and the data passed to it.
+
+ We don't expose constructors. To create you need to use helper functions
+ such as 'mkTestValidator' and 'mkTestMintingPolicy'. In case you intend
+ to test something tricky, you can use 'mkTestValidatorUnsafe'
+ and 'mkTestMintingPolicyUnsafe' to create a 'TestScript'
+ that accepts datum and/or redeemer inconsistent with its internal type.
+
+ @since 6.0
+-}
+data TestScript (p :: Purpose) where
+  -- | since 6.0
+  TestValidator ::
+    forall (d :: Type) (r :: Type).
+    {getTestValidator :: Validator} ->
+    TestScript ( 'ForSpending d r)
+  -- | since 6.0
+  TestMintingPolicy ::
+    forall (r :: Type).
+    {getTestMintingPolicy :: MintingPolicy} ->
+    TestScript ( 'ForMinting r)
+
+-- | @since 6.0
+deriving stock instance Show (TestScript p)
+
+{- | A wrapper for an untyped 'Validator'. This is similar
+ to 'WrappedValidatorType' from the Plutus 'Ledger.Typed.Scripts' module.
+ It`s aim is to force you to use 'toTestValidator' to create a 'TestValidator'.
+
+ @since 6.0
+-}
+newtype WrappedValidator
+  = WrappedValidator (BuiltinData -> BuiltinData -> BuiltinData -> ())
+
+{-# INLINEABLE getWrappedValidator #-}
+getWrappedValidator ::
+  WrappedValidator -> (BuiltinData -> BuiltinData -> BuiltinData -> ())
+getWrappedValidator (WrappedValidator v) = v
+
+{- | A wrapper for an untyped 'MintingPolicy'. This is similar
+ to 'WrappedMintingPolicyType' from the Plutus 'Ledger.Typed.Scripts' module.
+ It`s aim is to force you to use 'toTestMintingPolicy' to create
+ a 'TestMintingPolicy'.
+
+ @since 6.0
+-}
+newtype WrappedMintingPolicy
+  = WrappedMintingPolicy (BuiltinData -> BuiltinData -> ())
+
+{-# INLINEABLE getWrappedMintingPolicy #-}
+getWrappedMintingPolicy ::
+  WrappedMintingPolicy -> (BuiltinData -> BuiltinData -> ())
+getWrappedMintingPolicy (WrappedMintingPolicy mp) = mp
+
+{- | 'mkTestValidator' provide the way for creating 'TestScript' with
+ the given 'Validator', parameterized with proper datum and redeemer types.
+
+  = Usage
+
+ > testValidator :: TestScript ( 'ForSpending SomeType SomeOtherType)
+ > testValidator =
+ >  mkTestValidator
+ >    $$(compile [||myValidator||])
+ >    $$(compile [||toTestValidator||])
+
+ @since 6.0
+-}
+mkTestValidator ::
+  forall (datum :: Type) (redeemer :: Type) (ctx :: Type).
+  CompiledCode (datum -> redeemer -> ctx -> Bool) ->
+  CompiledCode ((datum -> redeemer -> ctx -> Bool) -> WrappedValidator) ->
+  TestScript ( 'ForSpending datum redeemer)
+mkTestValidator v wr =
+  TestValidator $
+    mkValidatorScript $
+      applyCode $$(compile [||getWrappedValidator||]) $
+        wr `applyCode` v
+
+{- | 'mkTestValidatorUnsafe' provide the way for creating 'TestScript' with
+ the given 'Validator', parameterized with any datum and redeemer types.
+ You can use it in case you going to test something tricky and need to break
+ types compatibility.
```

Comment:
Better phrasing:

As 'mkTestValidator', but without any guarantees of consistency regarding datum and redeemer types. You can use this in case you need to test something tricky, and the 'mkTestValidator' restrictions make this impossible.

I would also add a note that this is not safe.